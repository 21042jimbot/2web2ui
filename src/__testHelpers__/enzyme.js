export const findButton = (wrapper, buttonLabel) =>
  wrapper.findWhere(
    (node) => node.type() === 'button' && new RegExp(buttonLabel).test(node.text())
  );
