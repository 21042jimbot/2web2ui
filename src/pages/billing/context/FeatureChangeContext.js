import React, { createContext, useState, useContext, useEffect, useMemo } from 'react';
import _ from 'lodash';
import { connect } from 'react-redux';

import { getSubscription } from 'src/actions/billing';
import { selectPlansByKey } from 'src/selectors/accountBillingInfo';
// import { pluralString } from 'src/helpers/string';

const FeatureChangeContext = createContext({});

export const FeatureChangeProvider = ({
  children,
  plans,
  subscription,
  selectedBundle,
  loading,
  getSubscription,
}) => {
  const [actions, updateActions] = useState({});
  useEffect(() => {
    getSubscription();
  }, [getSubscription]);

  // TODO: Used to recheck subscription on refocusing tab. Doesn't matter now since there's no products that can be checked
  // //Rechecks conditions on re-entering tab. Only initializes once
  // const checkConditions = useCallback(() => {
  //   getSubscription();
  // }, [getSubscription]);
  // useEffect(() => {
  //   window.addEventListener('focus', checkConditions);
  //   return () => {
  //     window.removeEventListener('focus', checkConditions);
  //   };
  // }, [checkConditions]);

  //Keys the selected plan by product to make for easier comparison
  const selectedPlansByProduct = useMemo(() => {
    const { products } = selectedBundle;
    return products.reduce(
      (res, { product, plan: planId }) => ({
        ...res,
        [product]: plans[planId],
      }),
      {},
    );
  }, [plans, selectedBundle]);

  // Used for finding the features that need to have a proper function
  // Inserts into actions if it's got a conflicting issue
  // Updates if it was already in actions had conflicting issue
  const calculateDifferences = () => {
    if (!subscription) {
      return;
    }

    const { products: currentProducts } = subscription;

    const diffObject = currentProducts.reduce((resObject, { product }) => {
      const comparedPlan = selectedPlansByProduct[product];

      switch (product) {
        case 'dedicated_ip':
          if (actions.dedicated_ip || !comparedPlan) {
            resObject[product] = {
              label: 'Dedicated IPs',
              description: (
                <div>
                  <span>
                    {`Your new plan doesn't include dedicated IPs. 
                    Your current IP(s) will be removed at the end of your current billing cycle.`}
                  </span>
                </div>
              ),
            };
          }
          return resObject;
        case 'sso':
        case 'tfa_required':
          if (actions.auth || !comparedPlan) {
            resObject.auth = {
              label: 'Authentication and Security',
              description:
                'Your new plan no longer allows for single sign-on and account-wide requirement of two-factor authentication.',
            };
          }
          return resObject;
        case 'messaging':
        default:
          return resObject;
      }
    }, {});
    updateActions({ ...actions, ...diffObject });
  };

  useMemo(calculateDifferences, [subscription]);

  //Evaluates condition and generates action if condition exists
  const featuresWithActions = useMemo(
    () =>
      _.map(actions, (action, key) => ({
        ...action,
        key,
      })),
    [actions],
  );

  //Checks if all provided conditions are good
  const value = {
    features: featuresWithActions,
    loading,
  };

  return <FeatureChangeContext.Provider value={value}>{children}</FeatureChangeContext.Provider>;
};

const mapStateToProps = state => ({
  plans: selectPlansByKey(state),
  subscription: state.billing.subscription,
  loading: state.billing.loading,
});

export const FeatureChangeContextProvider = connect(mapStateToProps, { getSubscription })(
  FeatureChangeProvider,
);

export const useFeatureChangeContext = () => useContext(FeatureChangeContext);

export default FeatureChangeContext;
