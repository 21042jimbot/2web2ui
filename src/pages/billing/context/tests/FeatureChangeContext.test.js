import React from 'react';
import { shallow } from 'enzyme';
import { FeatureChangeProvider } from '../FeatureChangeContext';

const defaultProps = {
  plans: {
    '100K-premier-0519': {
      product: 'messaging',
      plan: 'plan-code',
    },
    'sso-0519': {
      product: 'sso',
      plan: 'sso-0519',
    },
    'subaccounts-0519': {
      product: 'subaccounts',
      plan: 'subaccounts-0519',
      limit: 15,
    },
    'subaccounts-starter': {
      product: 'subaccounts',
      plan: 'subaccounts-starter',
      limit: 0,
    },
    'tfa-required-0519': {
      product: 'tfa_required',
      plan: 'tfa-required-0519',
    },
    'ip-0519': {
      product: 'dedicated_ip',
      plan: 'ip-0519',
      volume: 1,
      limit: 4,
    },
  },
  subscription: {
    products: [
      {
        product: 'messaging',
        plan: '100K-premier-0519',
      },
      {
        product: 'subaccounts',
        plan: 'subaccounts-0519',
        quantity: 10,
        limit: 15,
      },
      {
        product: 'sso',
        plan: 'sso-0519',
      },
      {
        product: 'tfa_required',
        plan: 'tfa-required-0519',
      },
      {
        product: 'dedicated_ip',
        plan: 'ip-0519',
      },
    ],
  },
  selectedBundle: {
    products: [
      { product: 'messaging', plan: '100K-premier-0519' },
      { product: 'sso', plan: 'sso-0519' },
      { product: 'tfa_required', plan: 'tfa-required-0519' },
      { product: 'subaccounts', plan: 'subaccounts-0519' },
      { product: 'dedicated_ip', plan: 'ip-0519' },
    ],
  },
  loading: false,
  getSubscription: jest.fn(),
};

const subject = (props = {}, children = 'child') =>
  shallow(
    <FeatureChangeProvider {...defaultProps} {...props}>
      {children}
    </FeatureChangeProvider>,
  );

describe('FeatureChangeContext', () => {
  it('should render correctly', () => {
    const wrapper = subject();
    expect(wrapper.prop('value')).toEqual(
      expect.objectContaining({
        features: [],
        loading: false,
      }),
    );
  });

  it('should pass loading state', () => {
    const wrapper = subject({ loading: true });
    expect(wrapper.prop('value')).toEqual(
      expect.objectContaining({
        loading: true,
      }),
    );
  });

  it('should render acknowledgement for a change in auth', () => {
    const wrapper = subject({
      selectedBundle: {
        products: [
          { product: 'messaging', plan: '100K-premier-0519' },
          { product: 'subaccounts', plan: 'subaccounts-0519' },
          { product: 'dedicated_ip', plan: 'ip-0519' },
        ],
      },
    });
    expect(wrapper.prop('value')).toMatchSnapshot();
  });

  it('should render acknowledgement for a change in limit on subaccounts', () => {
    const wrapper = subject({
      selectedBundle: {
        products: [
          { product: 'messaging', plan: '100K-premier-0519' },
          { product: 'subaccounts', plan: 'subaccounts-starter' },
          { product: 'dedicated_ip', plan: 'ip-0519' },
          { product: 'sso', plan: 'sso-0519' },
          { product: 'tfa_required', plan: 'tfa-required-0519' },
        ],
      },
    });
    expect(wrapper.prop('value')).toMatchSnapshot();
  });

  it('should render acknowledgement for a change in dedicated IPs', () => {
    const wrapper = subject({
      selectedBundle: {
        products: [
          { product: 'messaging', plan: '100K-premier-0519' },
          { product: 'sso', plan: 'sso-0519' },
          { product: 'tfa_required', plan: 'tfa-required-0519' },
          { product: 'subaccounts', plan: 'subaccounts-0519' },
        ],
      },
    });
    expect(wrapper.prop('value')).toMatchSnapshot();
  });
});
