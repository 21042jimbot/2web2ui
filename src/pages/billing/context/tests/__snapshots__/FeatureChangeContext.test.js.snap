// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FeatureChangeContext should render acknowledgement for a change in auth 1`] = `
Object {
  "features": Array [
    Object {
      "action": <Button
        onClick={[Function]}
        size="default"
      >
        Got it
      </Button>,
      "description": "Your new plan no longer allows for single sign-on and account-wide requirement of two-factor authentication.",
      "key": "auth",
      "label": "Authentication and Security",
      "value": undefined,
    },
  ],
  "isReady": false,
  "loading": false,
}
`;

exports[`FeatureChangeContext should render acknowledgement for a change in dedicated IPs 1`] = `
Object {
  "features": Array [
    Object {
      "action": <Button
        onClick={[Function]}
        size="default"
      >
        Got it
      </Button>,
      "description": <div>
        <span>
          Your new plan doesn't include dedicated IPs. 
                    Your current IP(s) will be removed at the end of your current billing cycle.
        </span>
      </div>,
      "key": "dedicated_ip",
      "label": "Dedicated IPs",
      "value": undefined,
    },
  ],
  "isReady": false,
  "loading": false,
}
`;
